OrbitCore/LinuxEventTracer.cpp

```diff
@@ -0,0 +1,126 @@
+#include "LinuxPerfUtils.h"
+#include "LinuxEventTracer.h"
+#include "LinuxEventTracerVisitor.h"
+#include "LinuxPerfEventProcessor.h"
+#include "LinuxUtils.h"
+#include "Capture.h"
+
+#include <vector>
+
+
+//-----------------------------------------------------------------------------
+void LinuxEventTracer::Start()
+{
+    PRINT_FUNC;
+#if __linux__
+    m_ExitRequested = false;
+
+    m_Thread = std::make_shared<std::thread>(&LinuxEventTracer::Run, &m_ExitRequested);
+
+    m_Thread->detach();
+#endif
+}
+
+//-----------------------------------------------------------------------------
+void LinuxEventTracer::Stop()
+{
+    m_ExitRequested = true;
+}
+
+void LinuxEventTracer::Run(bool* a_ExitRequested)
+{
+    // get the number of cpus
+    int cpus = std::stoi(LinuxUtils::ExecuteCommand("nproc"));
+
+    uint64_t tracepointID = LinuxUtils::GetTracePointID("sched", "sched_switch");
+    
+    std::vector<uint32_t> fds;
+
+    // open perf events for all cpus to keep track of process spawning
+    for (int cpu = 0; cpu < cpus; cpu++)
+    {
+        fds.push_back(LinuxPerfUtils::task_event_open(cpu)); 
+        fds.push_back(LinuxPerfUtils::tracepoint_event_open(tracepointID, -1 /*pid*/, cpu)); 
+    }
+
+    std::vector<std::shared_ptr<LinuxPerfRingBuffer>> ring_buffers;
+    for (int32_t fd : fds)
+    {
+        auto ring_buffer = std::make_shared<LinuxPerfRingBuffer>(fd);
+        ring_buffers.push_back(ring_buffer);
+    
+        // start recording via ioctl
+        LinuxPerfUtils::start_capturing(fd);
+    }
+
+    // keep threads in synch.
+    auto threads = LinuxUtils::ListThreads(Capture::GTargetProcess->GetID());
+    for (uint64_t tid : threads) {
+        Capture::GTargetProcess->AddThreadId(tid);
+    }
+
+    LinuxPerfEventProcessor event_buffer(std::make_unique<LinuxEventTracerVisitor>());
+
+    while (!(*a_ExitRequested))
+    {
+        usleep(5000);
+
+        // read from all ring buffers, create events and store them in the event_queue
+        for (std::shared_ptr<LinuxPerfRingBuffer> ring_buffer : ring_buffers)
+        {
+            // read everything that is new
+            while (ring_buffer->HasNewData()) {
+                perf_event_header header{};
+                ring_buffer->ReadHeader(&header);
+
+                // perf_event_header::type contains the type of record, e.g.,
+                //  PERF_RECORD_SAMPLE, PERF_RECORD_MMAP, etc., defined in enum
+                //  perf_event_type in perf_event.h.
+                if (header.type == PERF_RECORD_SAMPLE) {
+                    // we need to know the type of the perf_record_sample at compile time,
+                    //  i.e. for multiple perf_event_open calls, we need be able to
+                    //  distinguish the different ring buffers.
+                    auto sched_switch = ring_buffer->ConsumeRecord<perf_event_record<sched_switch_tp>>(header);
+                    auto e = std::make_shared<LinuxSchedSwitchEvent>(sched_switch);
```

Comment:
I really wanted to avoid direct memory management here, as I don't think it is needed.
But sure, I have to admit, that especially the shared_ptr comes with some overhead.
Would it be an option for you, to go for unique_ptrs instead of shared_ptrs? They are way more lightweight. This would only require some tweaking of the EventProcessor datastructure, that has a priority queue internally.

As we really need to be careful, to sort the events from all the different ring_buffers and process them in sync with their timestamps. So I can't just add them into a Blockchain class.
Maybe, I could add them there and only store the pointers in the queue, but then the blocks would get fragmented, as the events are not necessarily processed (and therefore "freeed" afterwards) in the order they were  allocated. So they might not been removed from the blockchain in the (reverse-) order as they were allocated.

Reply:
See https://github.com/pierricgimmig/orbitprofiler/pull/82#discussion_r348802270.

Reply:
Having a system-wide mode now, that directly converts the events into ContextSwitches and send them to the backend, where they were stored cache friendly.
Despite that, I use unique_ptrs instead of shared_ptrs in the event processor now. This should be way better in terms of overhead. Still not cache friendly though. However, did not noticed any issue so far.